modular/gater/migrate_handler.go

```diff
@@ -299,3 +301,261 @@ func (g *GateModular) getSwapOutApproval(w http.ResponseWriter, r *http.Request)
+		reqCtx.Context(), bucketID, bucketSize, quota.GetMonth())
+	if err != nil {
+		log.CtxErrorw(reqCtx.Context(), "failed to get bucket read quota", "bucket_id",
+			bucketID, "error", err)
+		return
+	}
+
+	bz, err := quota.Marshal()
+	if err != nil {
+		log.CtxErrorw(reqCtx.Context(), "failed to marshal", "bucket_id",
+			bucketID, "error", err)
+		return
+	}
+
+	w.Header().Set(GnfdSignedApprovalMsgHeader, hex.EncodeToString(bz))
+	w.Header().Set(ContentTypeHeader, ContentTypeXMLHeaderValue)
+
+	log.CtxInfow(reqCtx.Context(), "succeed to pre bucket migrate and deduct quota", "bucket_id",
+		bucketID, "quota", quota, "bucket_quota_size", bucketSize)
+}
+
+// postMigrateBucketHandler notifying the source sp about the completion of migration bucket
+func (g *GateModular) postMigrateBucketHandler(w http.ResponseWriter, r *http.Request) {
+	var (
+		err        error
+		reqCtx     *RequestContext
+		bucketID   uint64
+		postMsg    []byte
+		bucketSize uint64
+	)
+
+	defer func() {
+		reqCtx.Cancel()
+		if err != nil {
+			reqCtx.SetError(err)
+			reqCtx.SetHTTPCode(int(gfsperrors.MakeGfSpError(err).GetHttpStatusCode()))
+			MakeErrorResponse(w, gfsperrors.MakeGfSpError(err))
+		} else {
+			reqCtx.SetHTTPCode(http.StatusOK)
+		}
+		log.CtxDebugw(reqCtx.Context(), reqCtx.String())
+	}()
+
+	reqCtx, _ = NewRequestContext(r, g)
+	postMigrateBucketHeader := r.Header.Get(GnfdPostMigrateBucketMsgHeader)
+	if postMsg, err = hex.DecodeString(postMigrateBucketHeader); err != nil {
+		log.CtxErrorw(reqCtx.Context(), "failed to parse migrate swap out header", "error", err)
+		err = ErrDecodeMsg
+		return
+	}
+	postMigrateBucket := types.GfSpPostMigrateBucket{}
+	if err = json.Unmarshal(postMsg, &postMigrateBucket); err != nil {
+		log.CtxErrorw(reqCtx.Context(), "failed to unmarshal post migrate bucket msg", "error", err)
+		err = ErrDecodeMsg
+		return
+	}
+
+	bucketID = postMigrateBucket.GetBucketId()
+
+	quota, err := g.baseApp.GfSpClient().GetLatestBucketReadQuota(
+		reqCtx.Context(), bucketID)
+	if err != nil {
+		log.CtxErrorw(reqCtx.Context(), "failed to get bucket read quota", "bucket_id",
+			bucketID, "error", err)
+		return
+	}
+
+	if postMigrateBucket.GetFinished() {
+		bz, err := quota.Marshal()
+		if err != nil {
+			log.CtxErrorw(reqCtx.Context(), "failed to marshal", "bucket_id",
+				bucketID, "error", err)
+			return
+		}
+		w.Header().Set(GnfdSignedApprovalMsgHeader, hex.EncodeToString(bz))
+	} else {
+		// get bucket quota and check TODO month check
+		bucketSize, err = g.getBucketTotalSize(reqCtx.Context(), bucketID)
+		if err != nil {
+			return
+		}
+		migratedBytes := postMigrateBucket.GetMigratedBytesSize()
+		extraQuota := bucketSize - migratedBytes
+		if migratedBytes > bucketSize {
+			log.CtxErrorw(reqCtx.Context(), "failed to recoup extra quota to user", "error", err)
+			return
+		}
+		quotaUpdateErr := g.baseApp.GfSpClient().RecoupQuota(reqCtx.Context(), postMigrateBucket.GetBucketId(), extraQuota, quota.GetMonth())
```

Comment:
RecoupQuota should be called in source SP , not dest SP, check if it is right here